### PR TITLE
lib-hooks: document with comments

### PR DESCRIPTION
**Pull request by liam-fitzgerald**
documents both the helper libraries for building hooks.

### CODE REVIEW

File: pkg/arvo/lib/pull-hook.hoon [hunk 1]
```diff
@@ -1,17 +1,49 @@
+::  lib/pull-hook/hoon: helper for creating a push hook
```
Comment (Fang-) [inline pkg/arvo/lib/pull-hook.hoon:1]:
```suggestion
::  lib/pull-hook: helper for creating a pull-hook
```

File: pkg/arvo/lib/pull-hook.hoon [hunk 2]
```diff
@@ -1,17 +1,49 @@
...
+::  
+::   lib/pull-hook is a helper for automatically pulling data from a
+::   corresponding push-hook to a store. 
+::   
+::   ## Interfacing notes:
+::
+::   lib/push-hook provides the inner door with its own separate state.
```
Comment (Fang-) [inline pkg/arvo/lib/pull-hook.hoon:8]: The phrasing here feels a bit confusing, but all it's really saying is that the inner agent's `+on-save` result gets threaded through to its `+on-load` transparently, right? Might not even be worth mentioning, if it just preserves expected behavior. But ymmv.

File: pkg/arvo/lib/pull-hook.hoon [hunk 3]
```diff
@@ -1,17 +1,49 @@
...
+::   The inner door may interact with the library by producing cards or
+::   scrying. Do not pass any cards on a wire beginning with /helper as
```
Comment (Fang-) [inline pkg/arvo/lib/pull-hook.hoon:10]: >or scrying

I don't see any special scry handling in this. If there was, you'd want to mention the prefix path used for these.

But also, this seems potentially problematic. At the very least you want to pass it onto the underlying agent, right?

https://github.com/urbit/urbit/blob/d91de5f2769060e81bd46c696f373064a9bf9033/pkg/arvo/lib/pull-hook.hoon#L246

File: pkg/arvo/lib/pull-hook.hoon [hunk 4]
```diff
@@ -1,17 +1,49 @@
...
 ::
 |%
 +$  card   card:agent:gall
 ::
+::  $config: configuration for the pull hook
+::
+::    .store-name: name of the store to proxy subscriptions to.
+::    .update-mark: mark that updates will be tagged with
+::    .push-hook-name: name of the corresponding push-hook
+
```
Comment (Fang-) [inline pkg/arvo/lib/pull-hook.hoon:34]:
```suggestion
::
```

File: pkg/arvo/lib/pull-hook.hoon [hunk 5]
```diff
@@ -1,17 +1,49 @@
...
 /+  default-agent, resource
 ::
 ::
 |%
 +$  card   card:agent:gall
 ::
+::  $config: configuration for the pull hook
+::
+::    .store-name: name of the store to proxy subscriptions to.
+::    .update-mark: mark that updates will be tagged with
```
Comment (Fang-) [inline pkg/arvo/lib/pull-hook.hoon:32]: Suggestion:

```suggestion
::    .store-name: name of the store to send subscription updates to
::    .update-mark: mark that updates will be tagged with
```

File: pkg/arvo/lib/pull-hook.hoon [hunk 6]
```diff
@@ -77,7 +109,9 @@
     *[(list card) _^|(..on-init)]
   ::  +on-pull-nack: handle failed pull subscription
```
Comment (Fang-) [inline pkg/arvo/lib/pull-hook.hoon:110]: I'd recommend moving these specialized arms to be _above_ the standard gall ones, so you can find them more quickly when reading top-to-bottom.

File: pkg/arvo/lib/push-hook.hoon [hunk 7]
```diff
@@ -1,15 +1,55 @@
+::  lib/push-hook/hoon: helper for creating a push hook
```
Comment (Fang-) [inline pkg/arvo/lib/push-hook.hoon:1]:
```suggestion
::  lib/push-hook: helper for creating a push-hook
```

File: pkg/arvo/lib/push-hook.hoon [hunk 8]
```diff
@@ -1,15 +1,55 @@
...
+::  
+::   lib/push-hook is a helper for automatically pushing data from a
+::   local store to the corresponding pull-hook on remote ships. It also
+::   proxies remote pokes to the store.
```
Comment (Fang-) [inline pkg/arvo/lib/push-hook.hoon:5]: > automatically pushing data from a local store to ... remote ships

This is why I don't like the "pull-push" phrasing. You're not so much _pushing_ as you are answering requests. Am I alone in thinking this is weird?

File: pkg/arvo/lib/push-hook.hoon [hunk 9]
```diff
@@ -1,15 +1,55 @@
...
+::
+::   ## Interfacing notes:
+::
+::   lib/push-hook provides the inner door with its own separate state.
+::   The inner door may interact with the library by producing cards or
+::   scrying. Do not pass any cards on a wire beginning with /helper as
```
Comment (Fang-) [inline pkg/arvo/lib/push-hook.hoon:11]: There are no special scries here, either. But at least it passes `+on-peek` calls through to `og` properly!

File: pkg/arvo/lib/pull-hook.hoon (hunk 3 above)
Comment (Fang-) [reply pkg/arvo/lib/pull-hook.hoon:10]: (Also `/helper` is a bit unhelpful as far as descriptive prefixes go. It's the generic name, but you might want to have multiple helper libraries that do different things, etc. etc.)

File: pkg/arvo/lib/push-hook.hoon [hunk 10]
```diff
@@ -1,15 +1,55 @@
...
+::   these wires are reserved by this library. Any watches/pokes/peeks
+::   not listed below will be routed to the inner door.
+::
+::   ##  Subscription paths
+::
+::   /resource/[resource]: Receive initial state and updates to
+::   .resource. .resource should be encoded with de-path:resource from
```
Comment (Fang-) [inline pkg/arvo/lib/push-hook.hoon:18]:
```suggestion
::   .resource. .resource should be encoded with en-path:resource from
```

File: pkg/arvo/lib/push-hook.hoon (hunk 8 above)
Comment (liam-fitzgerald) [reply pkg/arvo/lib/push-hook.hoon:5]: I do think the listen/proxy dichotomy is clearer, but pull/push is clearer.

File: pkg/arvo/lib/pull-hook.hoon (hunk 3 above)
Comment (liam-fitzgerald) [reply pkg/arvo/lib/pull-hook.hoon:10]: Oh the plan was to have `/helper` be generally a reserved prefix for helper libraries. i.e. pull-hook uses /helper/pull-hook and push-hook uses /helper/push-hook as their wire prefixes.

File: pkg/arvo/lib/push-hook.hoon (hunk 8 above)
Comment (tacryt-socryp) [reply pkg/arvo/lib/push-hook.hoon:5]: I like having symmetrical, short names. Lmk if you have recommendations

Comment (Fang-) [reply pkg/arvo/lib/push-hook.hoon:5]: Symmetrical in what sense? Character length, or semantics?

I think that's less important than being clear about what these things do. Best alternative I got is hear/show.

**Comment by matildepark** [comment]
@Fang- anything left here?

**Comment by Fang-** [comment]
The comments still say the wrapper agents can be interacted with by scrying, even though they don't do any scry handling themselves. That should be the last thing left though!

**Comment by matildepark** [comment]
@liam-fitzgerald It's been a month for a comment change, let's get this sweeped up

**Pull request merged**